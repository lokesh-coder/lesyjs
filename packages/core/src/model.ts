import { Chalk } from "chalk";
import { Ora } from "ora";

export interface SimpleObj<T> {
  [key: string]: T;
}

export interface ArgSchema {
  required: boolean;
}

export interface FlagSchema {
  required: boolean;
  aliases: string[];
}

export const enum MiddlewarePlacement {
  INIT = "INIT",
  START = "START",
  PRE_VALIDATE = "PRE_VALIDATE",
  POST_VALIDATE = "POST_VALIDATE",
  PRE_PARSE = "PRE_PARSE",
  POST_PARSE = "POST_PARSE",
  PRE_RUN = "PRE_RUN",
  POST_RUN = "POST_RUN",
  END = "END",
}

export type MiddlewareFn = (
  context: Partial<State>,
  next?: () => any,
) => Partial<State>;

export interface Middleware {
  on: MiddlewarePlacement;
  run: MiddlewareFn;
  source?: string;
}

export type MiddlewareContext = Record<string, any>;

export type MiddlewarePlacements = Record<MiddlewarePlacement, Middleware[]>;

export interface CommandRunContext {
  args: { [key: string]: string };
  flags: { [key: string]: string };
  rest: { [key: string]: string };
  root: string;
  config: { [key: string]: any };
  feature: { [key: string]: any };
  utils: { color: Function; spinner: Function };
  argv: string[];
  command: Command;
}

export type ValidatorFn = (name: string, rule: any) => boolean;
export interface Validator {
  name: string;
  fn: ValidatorFn;
}
export interface Request {
  runCommand: (argv: string[]) => any;
  getCommandById: (ID: number) => Command;
  getCommandByName: (name: string) => Command;
  getCommands: () => Command[];
}

export type Utils = () => { color: Chalk; spinner: Ora };

interface DefaultConfig {
  defaultCommand: string;
}

export interface State {
  argv: string[];
  id: number;
  args: Record<string, string | number | boolean>;
  flags: Record<string, string | number | boolean>;
  rawArgs: string[];
  rawFlags: Record<string, string | boolean>;
  unknownArgs: string[];
  unknownFlags: string[];
  config: Record<string, any> & Partial<DefaultConfig>;
  root: string;
  feature: Record<string, any>;
  utils: Readonly<Partial<Utils>>;
  request: Partial<Request>;
  cmdRunError: Error;
  validators: Validator[];
  allCommands: Command[];
  runningCommand: Command;
}

export interface Command {
  id?: Readonly<number>;
  name: string;
  run: (ctx: Partial<State>) => any;
  description?: string;
  aliases?: string[];
  args?: any;
  flags?: any;
  group?: string;
  isVisible?: boolean;
  main?: string | null;
  src?: Readonly<string>;
}

export interface Validation {
  isPassed: boolean;
  errors: any[];
}

export interface FeatureObj {
  [key: string]: any;
}

export type Feature = (feature: FeatureObj, opts: any) => void;

export interface Constructor<T> extends Function {
  new (...args: any[]): T;
}
type GenericClass = { new (...args: any[]): any };
type CommandClass = { new (...args: any[]): Command };

type ExtractFunctionCommand<T> = Function & { prototype: T };
type ExtractStringCommand<T> = T extends string ? string : never;
type ExtractRawCommand<T> = T extends Command ? Command : never;
type ExtractClassCommand<T> = Function & { prototype: T };

export type CommandSource =
  | ExtractFunctionCommand<Function>
  | ExtractStringCommand<string>
  | ExtractRawCommand<Command>
  | ExtractClassCommand<Command>;

export type Plugin = string | [string, any] | any;
// export type Plugin<T> = T extends string ? string : [string, any];

// namespace Validation {}
// namespace Middleware {}
// namespace Feature {}
