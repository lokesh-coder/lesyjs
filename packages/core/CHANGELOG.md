# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

# [1.0.0-beta.13](https://github.com/lokesh-coder/lesyjs/compare/v1.0.0-beta.12...v1.0.0-beta.13) (2020-10-27)

**Note:** Version bump only for package @lesy/core





# [1.0.0-beta.12](https://github.com/lokesh-coder/lesyjs/compare/v1.0.0-beta.11...v1.0.0-beta.12) (2020-10-25)

**Note:** Version bump only for package @lesy/core





# [1.0.0-beta.11](https://github.com/lokesh-coder/lesyjs/compare/v1.0.0-beta.10...v1.0.0-beta.11) (2020-10-14)

**Note:** Version bump only for package @lesy/core





# [1.0.0-beta.10](https://github.com/lokesh-coder/lesyjs/compare/v1.0.0-beta.9...v1.0.0-beta.10) (2020-09-27)


### Bug Fixes

* **core:** when command prop value is boolean is ignore in command ([fa79825](https://github.com/lokesh-coder/lesyjs/commit/fa79825f07c3e93f987cb6f804edac7cbc666ad7))





# [1.0.0-beta.9](https://github.com/lokesh-coder/lesyjs/compare/v1.0.0-beta.8...v1.0.0-beta.9) (2020-09-16)

**Note:** Version bump only for package @lesy/core





# [1.0.0-beta.8](https://github.com/lokesh-coder/lesyjs/compare/v1.0.0-beta.7...v1.0.0-beta.8) (2020-09-16)


### Performance Improvements

* improve overall performance ([f263587](https://github.com/lokesh-coder/lesyjs/commit/f2635870954e202ddcfffc252c9dac4387abe3f1))





# [1.0.0-beta.7](https://github.com/lokesh-coder/lesyjs/compare/v1.0.0-beta.6...v1.0.0-beta.7) (2020-08-24)

**Note:** Version bump only for package @lesy/core





# [1.0.0-beta.6](https://github.com/lokesh-coder/lesyjs/compare/v1.0.0-beta.5...v1.0.0-beta.6) (2020-08-19)


### Bug Fixes

* **core:** set default cmd and ignore files prefix with _ ([fe9f781](https://github.com/lokesh-coder/lesyjs/commit/fe9f781ca6c3a04ebb77b63cbaad83b13c22c9b5))


### Features

* **core:** add support to ignore files while providing dir ([ff6b17f](https://github.com/lokesh-coder/lesyjs/commit/ff6b17f610e4229db43ac1d0ea19f09674c01fba))





# [1.0.0-beta.5](https://github.com/lokesh-coder/lesyjs/compare/v1.0.0-beta.4...v1.0.0-beta.5) (2020-06-22)

**Note:** Version bump only for package @lesy/core





# [1.0.0-beta.4](https://github.com/lokesh-coder/lesyjs/compare/v1.0.0-beta.3...v1.0.0-beta.4) (2020-06-21)

**Note:** Version bump only for package @lesy/core





# [1.0.0-beta.3](https://github.com/lokesh-coder/lesyjs/compare/v1.0.0-beta.2...v1.0.0-beta.3) (2020-06-17)

**Note:** Version bump only for package @lesy/core





# [1.0.0-beta.2](https://github.com/lokesh-coder/lesyjs/compare/v1.0.0-beta.1...v1.0.0-beta.2) (2020-06-09)

**Note:** Version bump only for package @lesy/core





# 1.0.0-beta.1 (2020-06-09)


### Features

* initial commit ([589700b](https://github.com/lokesh-coder/lesyjs/commit/589700ba0d0f738bbfd77c2f921b81c4098adec9))
