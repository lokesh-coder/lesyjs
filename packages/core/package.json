{
  "name": "@lesy/core",
  "version": "1.0.0-beta.9",
  "description": "",
  "main": "dist/index.js",
  "scripts": {
    "clean": "rm -rf ./dist",
    "compile": "npm run clean && tsc --build tsconfig.build.json",
    "build": "npm run clean && ncc build src/index.ts  -o dist -e @lesy/validator -e tsconfig-paths -e ts-node -e chalk -e ora -e argh -e trough --no-source-map-register -m"
  },
  "keywords": [
    "core",
    "cli",
    "lesy",
    "commands",
    "typescript"
  ],
  "author": {
    "email": "mexican.dirtyfellow@gmail.com",
    "name": "Lokesh Rajendran"
  },
  "license": "MIT",
  "dependencies": {
    "@lesy/test-utils": "^1.0.0-beta.9",
    "@lesy/validator": "^1.0.0-beta.9",
    "argh": "^1.0.0",
    "chalk": "^4.0.0",
    "ora": "^4.0.4",
    "trough": "1.0.5"
  },
  "publishConfig": {
    "access": "public"
  }
}
