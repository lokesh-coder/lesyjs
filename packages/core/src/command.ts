import { Command, Validation, State } from "./model";

class LesyCommand {
  private commands: Command[] = [];
  private aliasesDict: Object = {};
  private defaultProps: Command = {
    id: null,
    name: "",
    args: {},
    flags: {},
    aliases: [],
    main: null,
    group: "Commands",
    description: "",
    isVisible: true,
    run: () => {},
  };

  private static lastAddedID: number;

  constructor() {
    LesyCommand.lastAddedID = 0;
  }

  add(rawCmd: Command | Function, src: string) {
    const cmd = this.transformRawCommand(rawCmd);
    // tslint:disable-next-line: no-increment-decrement
    Object.assign(cmd, { src, id: LesyCommand.lastAddedID++ });
    this.commands.push(cmd);
    cmd.aliases.forEach((alias: any) => {
      this.aliasesDict[`${cmd.main || "root"}.${alias}`] = cmd.name;
    });
  }

  findCommand(
    args: string[] = [],
    flags = {},
    ancestor: string = "root",
  ): Partial<State> {
    let cmd: Command;
    const parent = this.aliasesDict[`${ancestor}.${args[0]}`];
    cmd = this.commands.find((cmd: Command) => cmd.name === parent);
    if (!parent) {
      console.log(
        `command \x1b[31m ${args[0] || "<empty>"} \x1b[0m not found!`,
      );
      process.exit(1);
    }

    const child = this.aliasesDict[`${parent}.${args[1]}`];

    if (child) return this.findCommand(args.splice(1), flags, parent);
    const resolvedArgs = this.resolveArgs(cmd, args.splice(1));
    const resolvedFlags = this.resolveFlags(cmd, flags);

    return {
      runningCommand: cmd,
      ...resolvedArgs,
      ...resolvedFlags,
    };
  }

  private resolveArgs(cmd: Command, rawArgs: string[]): Partial<State> {
    const args = {};
    const argSchemas = Object.keys(cmd.args);
    argSchemas.forEach((arg, i) => {
      args[arg] = rawArgs[i];
    });
    return {
      args,
      unknownArgs: rawArgs.slice(argSchemas.length),
    };
  }

  private resolveFlags(
    cmd: Command,
    flags: Record<string, string | boolean>,
  ): Partial<State> {
    const flattenFlags = Object.keys(cmd.flags).map((k: any) => [
      k,
      ...(cmd.flags[k].aliases || []),
    ]);
    const resolvedFlags = {};
    for (const key in flags) {
      flattenFlags.forEach((fflags: string[]) => {
        if (fflags.includes(key)) {
          fflags.forEach((f: string) => (resolvedFlags[f] = flags[key]));
        }
      });
    }
    return { flags: resolvedFlags };
  }

  getCommands(): Command[] {
    return this.commands;
  }

  getCommandById(id: number): Command {
    return this.commands.find((c: Command) => c.id === id);
  }

  getCommandByName(names: string[]): any {
    return this.findCommand(names).runningCommand;
  }

  async validate(
    command: Command,
    args: object,
    validators = [],
  ): Promise<Validation> {
    const { LesyValidator } = require("@lesy/validator");
    const schema = new LesyValidator(command.args);
    validators.forEach(({ name, fn }) => schema.register(name, fn));
    const validation = await schema.validate(args);
    const isPassed = validation.every((v: any) => v.passed);
    return { isPassed, errors: validation };
  }

  private formatNames(cmd: Command) {
    cmd.name = cmd.name || "default";
    cmd.aliases = [cmd.name, ...cmd.aliases].map(this.normalizeStr);
    cmd.name = cmd.aliases[0];
    return cmd;
  }

  private transformRawCommand(cmd: Command | Function): Command {
    let cmdObj: Command;
    if (typeof cmd === "object") {
      cmdObj = { ...this.defaultProps, ...cmd };
    } else if (this.isClass(cmd)) {
      cmdObj = new (cmd as any)() as Command;
      Object.keys(this.defaultProps).forEach((p: any) => {
        if (cmdObj[p] != null) return;
        Object.defineProperty(cmdObj, p, {
          value: this.defaultProps[p],
          configurable: true,
          enumerable: true,
          writable: true,
        });
      });
    } else {
      cmdObj = { ...this.defaultProps, ...cmd };
      cmd(cmdObj);
    }

    return this.formatNames(cmdObj);
  }

  private isClass(funcOrClass: Function): boolean {
    return (
      typeof funcOrClass === "function" &&
      /^class\s/.test(Function.prototype.toString.call(funcOrClass))
    );
  }

  private normalizeStr(name: string): string {
    return name.replace(/[^\w]/gi, "-");
  }
}

export { LesyCommand };
