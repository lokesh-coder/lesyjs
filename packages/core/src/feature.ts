import { Feature, FeatureObj } from "./model";
import { utils } from "./utilities";

class LesyFeature {
  private feature = {};
  constructor(private root: string) {}

  add(featureFn: Feature) {
    featureFn(this.feature, this.root, utils);
  }
  getFeatures(): FeatureObj {
    return this.feature;
  }
}
export { LesyFeature };
