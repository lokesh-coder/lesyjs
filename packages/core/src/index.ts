export { LesyCore, LesyCoreClass } from "./core";
export * from "./model";
