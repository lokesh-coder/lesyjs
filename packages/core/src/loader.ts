import { lstatSync, readdirSync, statSync } from "fs";
import { join, sep } from "path";

import { Command, Plugin } from "./model";
import { LesyCommand } from "./command";

class LesyLoader {
  private root: string;
  pluginConfigs = {};
  cmdCtrl: any;
  mwCtrl: any;
  featCtrl: any;

  commands: any[] = [];
  middlewares: any[] = [];
  features: any[] = [];
  validators: any[] = [];

  constructor(
    {
      commands = [],
      features = [],
      middlewares = [],
      plugins = [],
      validators = [],
    }: {
      commands?: Command[] | string[];
      features?: string[];
      middlewares?: string[];
      validators?: any[];
      plugins?: Plugin[];
    },
    root: string,
  ) {
    this.root = root;
    const corePlugin = { commands, middlewares, features, validators };
    this.loadPlugins([corePlugin, ...plugins]);
  }

  loadFromObject(item: any, path: string, type: string) {
    this[`${type}Ctrl`].add(item, path);
  }
  loadFromFile(path: string, type: string) {
    const item = this.getModuleFromFile(path);
    this.loadFromObject(item, path, type);
  }
  loadFromDir(dir: string, type: string) {
    this.getFiles(dir).map((path: string) => this.loadFromFile(path, type));
  }
  load(items: any[], type: string) {
    items.forEach((item) => {
      if (typeof item !== "string") {
        return this.loadFromObject(item, "__OBJ__", type);
      }
      this.loadFilesAndDirs(item, type);
    });
  }

  loadPluginFromObject({
    commands = [],
    middlewares = [],
    features = [],
    validators = [],
  }): void {
    this.commands = this.commands.concat(commands);
    this.middlewares = this.middlewares.concat(middlewares);
    this.features = this.features.concat(features);
    this.validators = this.validators.concat(validators);
  }
  loadPluginFromFile(path: string): void {
    let plugin = this.getModuleFromFile(path);
    plugin = plugin["default"] || plugin;
    this.loadPluginFromObject(plugin);
  }
  loadPluginFromDir(dir: string): void {
    const pluginPath = require.resolve(dir);
    this.loadPluginFromFile(pluginPath);
  }
  loadPlugins(paths: Plugin[]): void {
    paths.forEach((pluginSource: any) => {
      if (Array.isArray(pluginSource)) {
        const [pluginPath, pluginConfig = {}] = pluginSource;
        this.pluginConfigs[pluginPath] = this.loadPluginsPlugins(pluginConfig);
        // tslint:disable-next-line: no-parameter-reassignment
        pluginSource = pluginPath;
      }
      try {
        const isObject = typeof pluginSource === "object";
        if (isObject) return this.loadPluginFromObject(pluginSource);
        const isFile = lstatSync(pluginSource).isFile();
        if (isFile) return this.loadPluginFromFile(pluginSource);
        const isDir = lstatSync(pluginSource).isDirectory();
        if (isDir) {
          return this.getDirectories(pluginSource).map((dir: string) =>
            this.loadPluginFromDir(dir),
          );
        }
      } catch (e) {}
      try {
        const pluginModule = require.resolve(pluginSource, {
          paths: [this.root],
        });
        this.loadPluginFromFile(pluginModule);
      } catch (e) {
        console.log(`${pluginSource} is not loaded`);
        console.log(`Error: ${e.message}`);
      }
    });

    this.cmdCtrl = new LesyCommand();
    this.load(this.commands, "cmd");
    if (this.features.length > 0) {
      const { LesyFeature } = require("./feature");
      this.featCtrl = new LesyFeature(this.root);
      this.load(this.features, "feat");
    }

    if (this.middlewares.length > 0) {
      const { LesyMiddleware } = require("./middleware");
      this.mwCtrl = new LesyMiddleware();
      this.load(this.middlewares, "mw");
    }
  }

  private formatPath(path: string): string {
    const isRemoteModule = path.includes("/node_modules/");
    const isJsProject = process.env.LESY_LANG === "js";
    return isRemoteModule || isJsProject ? path.replace(/\.ts$/, ".js") : path;
  }

  private getModuleFromFile(path: string): object {
    const formattedPath = this.formatPath(path);
    const module = require(formattedPath);
    return module.default || module;
  }

  private getDirectories(path: string): string[] {
    return readdirSync(path)
      .filter((fileOrDir: string) =>
        statSync(join(path, fileOrDir)).isDirectory(),
      )
      .map((dir: string) => join(path, dir));
  }

  private getFiles(path: string): string[] {
    const filePaths: string[] = [];

    const loop = (p: string): void => {
      readdirSync(p).forEach((file: string) => {
        const fullPath = join(p, file);
        if (lstatSync(fullPath).isDirectory()) {
          loop(fullPath);
        } else {
          if (this.isAllowedFile(file)) filePaths.push(fullPath);
        }
      });
    };
    loop(path);
    return filePaths;
  }

  private isAllowedFile(path: string) {
    const pathSegments = path.split(sep);
    const fileName = pathSegments[pathSegments.length - 1];
    if (fileName.startsWith("_")) return false;
    const ext = path.split(".");
    const lastIndex = ext.length - 1;
    if (process.env.LESY_LANG === "js") {
      return ext[lastIndex] === "js";
    }
    return ext[lastIndex] !== "map" && ext[lastIndex - 1] !== "d";
  }

  private loadFilesAndDirs(fileOrDirName: string, name: string): void {
    const fileOrDir = this.formatPath(fileOrDirName);
    const isFile = lstatSync(fileOrDir).isFile();
    if (isFile) return this.loadFromFile(fileOrDir, name);
    const isDir = lstatSync(fileOrDir).isDirectory();
    if (isDir) return this.loadFromDir(fileOrDir, name);
  }

  private loadPluginsPlugins(pluginConfig = {}) {
    const pluginsPlugins = pluginConfig["plugins"];
    if (!pluginsPlugins || !Array.isArray(pluginsPlugins)) return pluginConfig;
    const pluginsPluginsData = [];
    pluginsPlugins.forEach((pluginSource: string) => {
      try {
        const isFile = lstatSync(pluginSource).isFile();
        if (isFile) {
          pluginsPluginsData.push({
            src: pluginSource,
            module: this.getModuleFromFile(pluginSource),
          });
          return;
        }
      } catch (e) {}

      try {
        const paths = [this.root];
        pluginsPluginsData.push({
          src: pluginSource,
          module: this.getModuleFromFile(
            require.resolve(pluginSource, { paths }),
          ),
        });
      } catch (e) {
        console.log(`${pluginSource} is not loaded`);
        console.log(`Error: ${e.message}`);
      }
    });
    return {
      ...pluginConfig,
      plugins: pluginsPluginsData,
    };
  }
}

export { LesyLoader };
