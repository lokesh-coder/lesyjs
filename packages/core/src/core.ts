import argh from "argh";

import { LesyCommand } from "./command";
import { LesyFeature } from "./feature";
import { LesyLoader } from "./loader";
import { LesyMiddleware } from "./middleware";
import {
  MiddlewareContext,
  MiddlewarePlacement as $,
  Validator,
  State,
} from "./model";
import { utils } from "./utilities";

class LesyCoreClass {
  loader: LesyLoader;
  feature = {};
  root: string;
  config: Record<string, any>;
  validators: Validator[];

  private localState: State;
  private cmdCtrl!: LesyCommand;
  private mwCtrl!: LesyMiddleware;
  private featCtrl!: LesyFeature;

  async bootstrap({
    root,
    commands = [],
    middlewares = [],
    features = [],
    plugins = [],
    validators = [],
    config = {},
  }): Promise<LesyCoreClass> {
    this.root = root;
    this.validators = validators;
    this.loader = new LesyLoader(
      { commands, features, middlewares, plugins },
      this.root,
    );
    this.config = { ...config, ...this.loader.getPluginConfigs() };

    this.mwCtrl = this.loader.getMiddlewares();
    this.state = { root, utils, validators, config: this.config };
    await this.hook($.INIT);

    this.cmdCtrl = this.loader.getCommand();
    this.featCtrl = this.loader.getFeature();
    const allCmds = this.cmdCtrl.getCommands();
    this.feature = this.featCtrl.getFeatures();
    const request = this.getRequests();
    this.state = { request, feature: this.feature, allCommands: allCmds };
    await this.hook($.START);

    return this;
  }

  async run(argv: string[]): Promise<any> {
    this.state = { argv };

    const preParsedState = await this.hook($.PRE_PARSE);
    const { argv: rawArgs = [], ...rawFlags } = argh([...preParsedState.argv]);
    this.state = { rawArgs, rawFlags };

    this.state = await this.hook($.POST_PARSE);

    const { rawArgs: rArgs, rawFlags: rFlags } = this.state;
    const defaultCmdName = this.state.config.defaultCommand;
    const cmdNames = !rArgs.length ? [defaultCmdName] : rArgs;

    this.state = this.cmdCtrl.findCommand(cmdNames)(rFlags);

    this.state = await this.hook($.PRE_VALIDATE);

    const validation = await this.cmdCtrl.validate(
      this.state.runningCommand,
      this.state.args,
      this.state.validators,
    );
    validation.errors
      .filter((e: any) => !!e.error)
      .forEach((e: any) => console.error(e.error));
    if (!validation.isPassed) process.exit();

    this.state = await this.hook($.POST_VALIDATE); // dont need?

    this.state = await this.hook($.PRE_RUN);
    try {
      await this.state.runningCommand.run(this.state);
    } catch (e) {
      this.state = { cmdRunError: e };
    }

    await this.hook($.END);
  }

  private async hook(
    name: $,
    data: MiddlewareContext = this.state,
  ): Promise<MiddlewareContext> {
    return await this.mwCtrl.run(name, data);
  }

  private getRequests() {
    return {
      runCommand: this.run.bind(this),
      getCommandById: this.cmdCtrl.getCommandById.bind(this.cmdCtrl),
      getCommandByName: this.cmdCtrl.getCommandByName.bind(this.cmdCtrl),
      getCommands: this.cmdCtrl.getCommands.bind(this.cmdCtrl),
    };
  }

  private get state(): Partial<State> {
    return this.localState;
  }

  private set state(newState: Partial<State>) {
    const defaultState = {
      argv: [],
      id: null,
      args: {},
      flags: {},
      rawArgs: [],
      rawFlags: [],
      unknownArgs: [],
      unknownFlags: [],
      config: {},
      root: null,
      feature: {},
      utils: {},
      request: {},
      cmdRunError: null,
      validators: [],
      allCommands: null,
      runningCommand: null,
    };
    const currentState = this.localState || defaultState;
    this.localState = { ...currentState, ...newState } as State;
  }
}

const core = new LesyCoreClass();

export { core as LesyCore, LesyCoreClass };
