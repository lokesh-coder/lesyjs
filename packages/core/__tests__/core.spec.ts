import { LesyCoreClass } from "../src/core";

import { createFiles, deleteDir } from "@lesy/test-utils";
import { Command } from "../src/model";

describe("Core", () => {
  let data;
  let core: LesyCoreClass;
  let startMwMock;
  let initMwMock;
  let preparseMwMock;
  let postparseMwMock;
  let prevalidateMwMock;
  let prerunMwMock;
  let endMwMock;
  let cmdRunMock;
  beforeAll(() => {
    createFiles(
      {
        "plugin/package.json": `{"name":"plugin","main":"index.js"}`,
        "plugin/middleware.js": "module.exports={on:'END',run:()=>{}}",
        "plugin/feature.js": "module.exports=(x)=>(x.pluginft='abc')",
        "plugin/command.js": "module.exports={name:'plugincmd',run:()=>{}}",
        "plugin/index.js": `module.exports={
            commands:['${__dirname}/plugin/command.js'],
            features:['${__dirname}/plugin/feature.js'],
            middlewares:['${__dirname}/plugin/middleware.js'],
        }`,
      },
      __dirname,
    );
  });
  afterAll(() => {
    deleteDir([`${__dirname}/plugin`]);
  });
  beforeEach(() => {
    startMwMock = jest.fn((x: any) => x);
    initMwMock = jest.fn((x: any) => x);
    preparseMwMock = jest.fn((x: any) => x);
    postparseMwMock = jest.fn((x: any) => x);
    prevalidateMwMock = jest.fn((x: any) => x);
    prerunMwMock = jest.fn((x: any) => x);
    endMwMock = jest.fn((x: any) => x);
    cmdRunMock = jest.fn();

    data = {
      root: `${__dirname}`,
      plugins: [[`plugin`, { conf1: "xyz" }]],
      commands: [
        { name: "hello", run: () => {} },
        { name: "greet", args: { title: { required: true } }, run: () => {} },
        {
          name: "sample",
          args: { title: { required: true } },
          flags: { f1: {} },
          run: cmdRunMock,
        },
      ],
      middlewares: [
        { on: "START", run: startMwMock },
        { on: "INIT", run: initMwMock },
        { on: "PRE_PARSE", run: preparseMwMock },
        { on: "POST_PARSE", run: postparseMwMock },
        { on: "PRE_VALIDATE", run: prevalidateMwMock },
        { on: "PRE_RUN", run: prerunMwMock },
        { on: "END", run: endMwMock },
      ],
      features: [(x: any) => (x.mainft = "one")],
      config: { test: "xyz", defaultCommand: "plugincmd" },
    };
    core = new LesyCoreClass();
  });
  afterEach(() => {
    jest.clearAllMocks();
    startMwMock.mockClear();
    initMwMock.mockClear();
    preparseMwMock.mockClear();
    postparseMwMock.mockClear();
    prevalidateMwMock.mockClear();
    prerunMwMock.mockClear();
    endMwMock.mockClear();
    cmdRunMock.mockClear();
  });

  describe("bootstrap", () => {
    beforeEach(async () => {
      core = await core.bootstrap(data);
    });
    it("should set root property", () => {
      expect(core.root).toEqual(__dirname);
    });
    it("should get all commands", () => {
      expect(
        core["cmdCtrl"].getCommands().map((x: Command) => x.name),
      ).toEqual(["hello", "greet", "sample", "plugincmd"]);
    });
    it("should get all features", () => {
      expect(core["featCtrl"].getFeatures()).toEqual({
        mainft: "one",
        pluginft: "abc",
      });
    });
    it("should get all middlewares", () => {
      expect(core["mwCtrl"].get()).toEqual({
        END: [
          { on: "END", run: expect.any(Function) },
          { on: "END", run: expect.any(Function) },
        ],
        INIT: [{ on: "INIT", run: expect.any(Function) }],
        POST_PARSE: [{ on: "POST_PARSE", run: expect.any(Function) }],
        POST_RUN: [],
        POST_VALIDATE: [],
        PRE_PARSE: [{ on: "PRE_PARSE", run: expect.any(Function) }],
        PRE_RUN: [{ on: "PRE_RUN", run: expect.any(Function) }],
        PRE_VALIDATE: [{ on: "PRE_VALIDATE", run: expect.any(Function) }],
        START: [{ on: "START", run: expect.any(Function) }],
      });
    });
    it("should get config", () => {
      expect(core["config"]).toEqual({
        defaultCommand: "plugincmd",
        test: "xyz",
        plugin: {
          conf1: "xyz",
        },
      });
    });
    it("should get complete state", () => {
      core["state"].allCommands[3].src = "__DUMMY_SRC__";
      expect(core["state"]).toMatchSnapshot({
        root: expect.stringContaining("/packages/core/__tests__"),
      });
    });
    it("should call init middleware", () => {
      expect(initMwMock).toBeCalledWith(
        expect.objectContaining({ allCommands: null }),
      );
    });
    it("should call start middleware", () => {
      expect(startMwMock).toBeCalledWith(
        expect.objectContaining({ allCommands: core["state"].allCommands }),
      );
    });
  });

  describe("run", () => {
    beforeEach(async () => {
      core = await core.bootstrap(data);
      await core.run(["hello", "--name", "john"]);
    });
    it("should call PRE_PARSE hook", () => {
      expect(preparseMwMock).toBeCalledWith(
        expect.objectContaining({ argv: ["hello", "--name", "john"] }),
      );
    });
    it("should call POST_PARSE hook", () => {
      expect(postparseMwMock).toBeCalledWith(
        expect.objectContaining({
          rawArgs: ["hello"],
          rawFlags: { name: "john" },
        }),
      );
    });
    it("should run default command if no args provided", async () => {
      await core.run([]);
      expect(prevalidateMwMock).nthCalledWith(
        2,
        expect.objectContaining({
          runningCommand: expect.objectContaining({ name: "plugincmd" }),
        }),
      );
    });
    it("should validate args", async () => {
      const consoleSpy = jest.spyOn(console, "error");
      jest.spyOn(console, "error").mockImplementationOnce(() => {
        throw new Error("__PROCESS_EXIT__");
      });

      try {
        await core.run(["greet"]);
      } catch (error) {
        expect(consoleSpy).toBeCalledWith(
          "<required> validation failed for [title]",
        );
        expect(error.message).toBe("__PROCESS_EXIT__");
      }
    });
    it("should run PRE_RUN middleware", () => {
      expect(prerunMwMock).toBeCalledTimes(1);
    });
    it("should run END middleware", () => {
      expect(endMwMock).toBeCalledTimes(1);
    });
    it("should run command", async () => {
      await core.run(["sample", "xyz", "--f1", "qaz"]);
      expect(cmdRunMock).toBeCalledWith(
        expect.objectContaining({
          args: { title: "xyz" },
          flags: { f1: "qaz" },
        }),
      );
    });
    it("should capture error if command throws", async () => {
      cmdRunMock.mockImplementationOnce(() => {
        throw new Error("test error");
      });
      await core.run(["sample", "xyz", "--f1", "qaz"]);
      expect(core["state"].cmdRunError.message).toBe("test error");
    });
  });
});
