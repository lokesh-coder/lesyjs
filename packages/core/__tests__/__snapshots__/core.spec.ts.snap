// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Core bootstrap should get complete state 1`] = `
Object {
  "allCommands": Array [
    Object {
      "aliases": Array [
        "hello",
      ],
      "args": Object {},
      "description": "",
      "flags": Object {},
      "group": "Commands",
      "id": 0,
      "isVisible": true,
      "main": null,
      "name": "hello",
      "run": [Function],
      "src": "__OBJ__",
    },
    Object {
      "aliases": Array [
        "greet",
      ],
      "args": Object {
        "title": Object {
          "required": true,
        },
      },
      "description": "",
      "flags": Object {},
      "group": "Commands",
      "id": 1,
      "isVisible": true,
      "main": null,
      "name": "greet",
      "run": [Function],
      "src": "__OBJ__",
    },
    Object {
      "aliases": Array [
        "sample",
      ],
      "args": Object {
        "title": Object {
          "required": true,
        },
      },
      "description": "",
      "flags": Object {
        "f1": Object {},
      },
      "group": "Commands",
      "id": 2,
      "isVisible": true,
      "main": null,
      "name": "sample",
      "run": [MockFunction],
      "src": "__OBJ__",
    },
    Object {
      "aliases": Array [
        "plugincmd",
      ],
      "args": Object {},
      "description": "",
      "flags": Object {},
      "group": "Commands",
      "id": 3,
      "isVisible": true,
      "main": null,
      "name": "plugincmd",
      "run": [Function],
      "src": "__DUMMY_SRC__",
    },
  ],
  "config": Object {
    "defaultCommand": "plugincmd",
    "plugin": Object {
      "conf1": "xyz",
    },
    "test": "xyz",
  },
  "feature": Object {
    "mainft": "one",
    "pluginft": "abc",
  },
  "request": Object {
    "getCommandById": [Function],
    "getCommandByName": [Function],
    "getCommands": [Function],
    "runCommand": [Function],
  },
  "root": StringContaining "/packages/core/__tests__",
  "utils": Object {
    "color": [Function],
    "spinner": [Function],
  },
  "validators": Array [],
}
`;
